[inductor] test codegen with dynamic shapes

nkaretnikov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96934

Adds new tests that check for patterns in generated C++/Triton code to
see if it's dynamic.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

nkaretnikov committed:
[ci] test inductor codegen

[ghstack-poisoned]

nkaretnikov commented on .github/workflows/pull.yml:
```diff
@@ -1,7 +1,7 @@
 
 on:
-  pull_request:
+  # pull_request:
```
**Remove this!** (to avoid running other jobs during testing)

nkaretnikov commented on test/inductor/expect/CodegenCpuTests.test_upsample_nearest1d_cpu.expect:
```diff
@@ -0,0 +1,157 @@
+
+
+kernel_cpp_0 = async_compile.cpp('''
+#include "/tmp/torchinductor_nkaretnikov/ao/caomrguwaptybystzop6sbmdii67jqxwyhyzjizs5gbwosh2mdkg.h"
```
Will need to update these paths. Left them for testing purposes.

Will also need to add all other files. Waiting for CI.

nkaretnikov commented on test/inductor/test_torchinductor_codegen.py:
```diff
@@ -0,0 +1,123 @@
+from inductor.test_torchinductor import run_and_get_cpp_code, run_and_get_triton_code
+
+
+def check_codegen(
```
This is based on `check_model` and `check_model_cuda`

nkaretnikov commented on .ci/pytorch/test.sh:
```diff
@@ -255,6 +255,10 @@ test_inductor() {
 }
 
+test_inductor_codegen() {
+  python test/run_test.py --include inductor/test_torchinductor_codegen --verbose
```
Add some diffing between static and dynamic outputs?

nkaretnikov commented:
@ezyang Still WIP (see the comments above), seeking early feedback

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented on .github/workflows/unstable.yml:
```diff
@@ -40,8 +40,6 @@ jobs:
       test-matrix: |
         { include: [
-          { config: "distributed", shard: 1, num_shards: 2, runner: "linux.2xlarge" },
-          { config: "distributed", shard: 2, num_shards: 2, runner: "linux.2xlarge" },
```
**Add these two lines back!** (removed to avoid wasting compute during testing)

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented on test/inductor/expect/CodegenCpuTests.test_avg_pool2d_backward4_cpu.expect:
```diff
@@ -0,0 +1,51 @@
+async_compile.wait(globals())
+del async_compile
+
+def call(args):
```
No CPP kernel.

nkaretnikov commented on test/inductor/expect/CodegenCpuTests.test_pixel_shuffle_channels_last_cpu.expect:
```diff
@@ -0,0 +1,78 @@
+                tmp1.store(in_out_ptr0 + 8*i0);
+            }
+            #pragma omp for simd simdlen(4)
+            for(long i0=73728; i0<73728; i0+=1)
```
Unrelated to this PR, but this is an interesting loop condition.

nkaretnikov commented on test/inductor/expect/CodegenCpuTests.test_std_cpu.expect:
```diff
@@ -0,0 +1,544 @@
+                    tmp7_vec += tmp0;
+                }
+                #pragma omp for simd simdlen(4) reduction(+:tmp6) reduction(+:tmp7)
+                for(long i0=256; i0<256; i0+=1)
```
Same here and in a bunch of other places.

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented on .lintrunner.toml:
```diff
@@ -370,6 +370,7 @@ code = 'SPACES'
 include_patterns = ['**']
 exclude_patterns = [
     '**/contrib/**',
+    '**/*.expect',
```
CPP codegen emits code with tabs and trailing spaces.

ezyang commented:
This is too many expect tests. Which ones are you actually going to keep.

nkaretnikov commented:
Okay, new plan. For CPP, I'll check `for` loops for `ks`. For triton, I'll check that `xnumel` is never assigned to an integer literal (or something similar).

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented on test/inductor/test_torchinductor_codegen_dynamic_shapes.py:
```diff
@@ -99,7 +99,101 @@ def run(*ex, **kwargs):
 test_skips = {
+    #
+    # Failed to find dynamic for loop variable:
+    #
```
Some of these are "partially dynamic": some loop might be dynamic, some not.

nkaretnikov commented on test/inductor/test_torchinductor_codegen_dynamic_shapes.py:
```diff
@@ -99,7 +99,101 @@ def run(*ex, **kwargs):
+    "test_zeros_dynamic_shapes": ("cpu",),
+    #
+    # Failed to find for loop:
+    #
```
This includes code not having CPP code and CPP code that doesn't have any for loops.

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented:
@ezyang This is green. Thoughts on the current checks? CPP: `for` loop with integer literals, triton: `xnumel =` integer literal.

(If okay, I'll revert some WIP changes and enable the rest of CI.)

ezyang commented on test/inductor/test_torchinductor_codegen_dynamic_shapes.py:
```diff
@@ -0,0 +1,292 @@
+    torch._dynamo.reset()
+
+
+test_skips = {
```
Can we do xfails instead of skips?

ezyang commented:
I'm willing to give this a shot. cc @ngimel for any other opinions.

ngimel commented on test/inductor/test_torchinductor_codegen_dynamic_shapes.py:
```diff
@@ -0,0 +1,292 @@
+from inductor.test_torchinductor_dynamic_shapes import make_dynamic_cls
+
+
+def check_codegen(
```
a comment on what this is testing would be nice

ngimel commented on test/inductor/test_torchinductor_codegen_dynamic_shapes.py:
```diff
@@ -0,0 +1,292 @@
+            if "def triton" in line:
+                triton_kernel_found = True
+                continue
+            if "xnumel =" in line:
```
if there's no static numel, there's still no guarantee that the code is dynamic?

nkaretnikov replied:
That's the first idea I had. I feel like checking for anything else might be error-prone. Open to suggestions.

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented on test/inductor/test_torchinductor.py:
```diff
@@ -2137,7 +2137,7 @@ def run(*ex, **kwargs):
                 return mod(*ex, **kwargs)
 
             run = torch._dynamo.optimize(compile_fx_wrapper)(run)
-            code = run_and_get_cpp_code(run, (v,))
```
Here and elsewhere: changed the signature of `run_and_get_cpp_code` to match `run_and_get_triton_code`.

nkaretnikov commented:
@pytorchbot rebase -s

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 @EikanWang jgong5 @Guobing-Chen @XiaobingSuper zhuhaozhe blzheng @Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented:
@pytorchbot rebase -b master

nkaretnikov committed:
Update on "[ci] test inductor codegen"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

nkaretnikov commented:
Status update: as of this commit, this only fails codegen tests on unstable, which means this shouldn't be executed outside of the unstable label (as intended). I'll add the missing codegen skips/xfails and this should be ready.

nkaretnikov committed:
Update on "[ci] test inductor codegen"


- Adds new tests that check for patterns in generated C++/Triton code to
  see if it's dynamic
- Runs these tests as part of the unstable workflow
- Adds `--exclude-inductor-codegen-tests` to `test/run_test.py` to
  exclude these tests from other workflows
- Updates the `run_and_get_cpp_code` signature to match
  `run_and_get_triton_code`

[ghstack-poisoned]

nkaretnikov commented:
@pytorchbot rebase -b master

nkaretnikov committed:
Update on "[ci] test inductor codegen"


- Adds new tests that check for patterns in generated C++/Triton code to
  see if it's dynamic
- Runs these tests as part of the unstable workflow
- Adds `--exclude-inductor-codegen-tests` to `test/run_test.py` to
  exclude these tests from other workflows
- Updates the `run_and_get_cpp_code` signature to match
  `run_and_get_triton_code`

[ghstack-poisoned]

nkaretnikov commented:
Note: before this is merged, I'd like to review the skip list here. I've just noticed we're not skipping `test_arange5` because it passes, but it doesn't use `common`. So should be in the skip list. There might be more like it.

ezyang commented on .ci/pytorch/macos-test.sh:
```diff
@@ -46,7 +46,7 @@ test_python_shard() {
   setup_test_python
 
-  time python test/run_test.py --verbose --exclude-jit-executor --exclude-distributed-tests --shard "$1" "$NUM_TEST_SHARDS"
+  time python test/run_test.py --verbose --exclude-jit-executor --exclude-distributed-tests --exclude-inductor-codegen-tests --shard "$1" "$NUM_TEST_SHARDS"
```
Shouldn't we default these tests as excluded, if we expect them to be failing most of the time?

nkaretnikov replied:
If we decide to go with unstable, OK. But see my reply below. I think we should just enable this as part of ordinary workflows -- no CI changes would be necessary, too.

ezyang commented:
So what's our expected overall workflow once this lands? How are we keeping track of failures on the periodic flow?

nkaretnikov commented:
@ezyang

> So what's our expected overall workflow once this lands? How are we keeping track of failures on the periodic flow?

Look at unstable jobs on hud occasionally. Run with ciflow/unstable on certain PRs. Fix issues once in a while.

However, the original plan was to check the enitre codegen output, which is how we decided on unstable. Now, we only look for small substrings. Shouldn't be too flaky. We also have a way to skip tests. I propose we enable this by default. Do you agree?

nkaretnikov committed:
Update on "[ci] test inductor codegen"


- Adds new tests that check for patterns in generated C++/Triton code to
  see if it's dynamic
- Runs these tests as part of the unstable workflow
- Adds `--exclude-inductor-codegen-tests` to `test/run_test.py` to
  exclude these tests from other workflows

[ghstack-poisoned]

nkaretnikov commented on test/inductor/test_torchinductor_codegen_dynamic_shapes.py:
```diff
@@ -86,7 +86,7 @@ def run(*ex, **kwargs):
                 for_loop_found = True
                 self.assertTrue(
-                    "ks" in line,
+                    re.search(r";.*ks.*;", line) is not None,
```
Improved the dynamic var check a bit.

nkaretnikov committed:
Update on "[inductor] test codegen with dynamic shapes"


Adds new tests that check for patterns in generated C++/Triton code to
see if it's dynamic.

[ghstack-poisoned]

nkaretnikov commented:
Pushed an update that removes CI changes, so that this runs as part of ordinary workflows. (My comments above explain why.)

nkaretnikov committed:
Update on "[inductor] test codegen with dynamic shapes"


Adds new tests that check for patterns in generated C++/Triton code to
see if it's dynamic.

[ghstack-poisoned]

nkaretnikov committed:
Update on "[inductor] test codegen with dynamic shapes"


Adds new tests that check for patterns in generated C++/Triton code to
see if it's dynamic.

[ghstack-poisoned]

nkaretnikov commented:
@ezyang 3 minutes when I run locally:

```
pytest -rfxs test/inductor/test_torchinductor_codegen_dynamic_shapes.py  539.56s user 24.68s system 310% cpu 3:01.60 total
```

ezyang reviewed (approved):
OK, let's give a try, but yank it if it's causing people problems

nkaretnikov commented:
@pytorchbot rebase -b master

nkaretnikov committed:
Update on "[inductor] test codegen with dynamic shapes"


Adds new tests that check for patterns in generated C++/Triton code to
see if it's dynamic.

[ghstack-poisoned]

nkaretnikov commented:
@pytorchbot merge

nkaretnikov commented:
@pytorchbot merge
